File: 5 File Code/node.js/chat_server.js
const net = require('net');

const PORT = 9999;
const HOST = '127.0.0.1';

// Danh sách chứa các client đang online
let clients = [];

const server = net.createServer((socket) => {
    // Định danh client
    socket.name = `${socket.remoteAddress}:${socket.remotePort}`;
    
    // Thêm vào danh sách
    clients.push(socket);
    
    console.log(`[SERVER] ${socket.name} đã tham gia.`);
    broadcast(`==> ${socket.name} da tham gia phong chat!\n`, socket);

    // Xử lý khi nhận tin nhắn
    socket.on('data', (data) => {
        const message = data.toString().trim();
        // Kiểm tra nếu tin nhắn rỗng thì bỏ qua
        if (message.length === 0) return;

        console.log(`[LOG] ${socket.name}: ${message}`);
        
        // Gửi cho tất cả người khác
        broadcast(`Client ${socket.name} noi: ${message}\n`, socket);
    });

    // Xử lý khi client thoát
    socket.on('end', () => {
        clients.splice(clients.indexOf(socket), 1); // Xóa khỏi danh sách
        console.log(`[SERVER] ${socket.name} đã thoát.`);
        broadcast(`==> ${socket.name} da roi khoi phong chat.\n`, socket);
    });

    // Xử lý lỗi
    socket.on('error', (err) => {
        // Lỗi thường gặp là client bị tắt đột ngột (Connection reset)
        clients.splice(clients.indexOf(socket), 1);
        console.log(`[LỖI] ${socket.name} ngắt kết nối bất thường.`);
    });
});

// Hàm gửi tin nhắn cho tất cả TRỪ người gửi (sender)
function broadcast(message, sender) {
    clients.forEach((client) => {
        // Kiểm tra client còn sống và không phải là người gửi
        if (client !== sender && client.writable) {
            client.write(message);
        }
    });
}

server.listen(PORT, HOST, () => {
    console.log(`[NODEJS] Chat Server đang chạy trên port ${PORT}...`);
});

File: 5_file_code/golang/chat_server.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strings"
	"sync"
)

// Cấu hình
const (
	HOST = "localhost"
	PORT = "9999"
	TYPE = "tcp"
)

// Quản lý danh sách client an toàn với Mutex
var (
	clients = make(map[net.Conn]string) // Map lưu kết nối và tên
	mu      sync.Mutex                  // Khóa an toàn cho Map
)

func main() {
	listen, err := net.Listen(TYPE, HOST+":"+PORT)
	if err != nil {
		fmt.Println("Lỗi khởi tạo Server:", err)
		return
	}
	defer listen.Close()

	fmt.Println("[GO SERVER] Đang chạy trên " + HOST + ":" + PORT)
	fmt.Println("[GO SERVER] Sử dụng Goroutines để xử lý đa luồng...")

	for {
		conn, err := listen.Accept()
		if err != nil {
			fmt.Println("Lỗi chấp nhận kết nối:", err)
			continue
		}

		// --- TỐI ƯU HÓA TCP (Theo yêu cầu) ---
		// Ép kiểu về TCPConn để chỉnh options
		if tcpConn, ok := conn.(*net.TCPConn); ok {
			tcpConn.SetNoDelay(true) // Tắt Nagle Algorithm
			tcpConn.SetKeepAlive(true)
		}

		// Tạo một Goroutine riêng cho client này (Giống như phân thân)
		go handleClient(conn)
	}
}

func handleClient(conn net.Conn) {
	defer conn.Close()

	// Lấy địa chỉ client
	clientAddr := conn.RemoteAddr().String()
	fmt.Printf("[SERVER] Kết nối mới: %s\n", clientAddr)

	// Thêm vào danh sách (cần khóa Mutex)
	mu.Lock()
	clients[conn] = clientAddr
	mu.Unlock()

	// Gửi thông báo cho mọi người
	broadcastMessage(conn, "==> "+clientAddr+" đa tham gia phong chat!\n")

	reader := bufio.NewReader(conn)
	for {
		// Đọc tin nhắn (Chờ đến khi có ký tự xuống dòng)
		message, err := reader.ReadString('\n')
		if err != nil {
			// Client ngắt kết nối
			break
		}

		message = strings.TrimSpace(message)
		if len(message) > 0 {
			fmt.Printf("[%s]: %s\n", clientAddr, message)
			// Gửi tin nhắn cho các client khác
			broadcastMessage(conn, "Client "+clientAddr+" noi: "+message+"\n")
		}
	}

	// Xử lý khi client thoát
	mu.Lock()
	delete(clients, conn)
	mu.Unlock()

	fmt.Printf("[SERVER] %s đã thoát.\n", clientAddr)
	broadcastMessage(nil, "==> "+clientAddr+" đa roi khoi phong chat.\n")
}

// Hàm gửi tin nhắn cho tất cả (trừ người gửi)
func broadcastMessage(sender net.Conn, message string) {
	mu.Lock()
	defer mu.Unlock()

	for conn := range clients {
		if conn != sender {
			// Gửi tin nhắn
			conn.Write([]byte(message))
		}
	}
}
